packages/core/CalendarManager.ts

```diff
@@ -22,39 +23,99 @@ import type { EventResult } from "@calcom/types/EventManager";
 import getCalendarsEvents from "./getCalendarsEvents";
 
 const log = logger.getSubLogger({ prefix: ["CalendarManager"] });
+type CredentialForCalendarService<T> = T extends null
+  ? null
+  : T & {
+      delegatedTo: {
+        serviceAccountKey: {
+          client_email: string;
+          private_key: string;
+          client_id: string;
+        };
+      } | null;
+    };
 
-export const getCalendarCredentials = (credentials: Array<CredentialPayload>) => {
-  const calendarCredentials = getApps(credentials, true)
-    .filter((app) => app.type.endsWith("_calendar"))
-    .flatMap((app) => {
-      const credentials = app.credentials.flatMap((credential) => {
-        const calendar = getCalendar(credential);
-        return app.variant === "calendar" ? [{ integration: app, credential, calendar }] : [];
-      });
+/**
+ * CalendarService needs delegatedTo to have serviceAccountKey for DWD Credential. It fetches that.
+ */
+export async function getCredentialForCalendarService<
+  T extends ({ delegatedToId?: string | null } & Record<string, unknown>) | null
+>(credential: T): Promise<CredentialForCalendarService<T>> {
+  // Explicitly handle null case with type assertion
+  if (credential === null) return null as CredentialForCalendarService<T>;
 
-      return credentials.length ? credentials : [];
-    });
+  // When no delegatedToId, return with delegatedTo as null
+  if (!credential.delegatedToId) {
+    return {
+      ...credential,
+      delegatedTo: null,
+    } as CredentialForCalendarService<T>;
+  }
+  const domainWideDelegationRepository = new DomainWideDelegationRepository();
+  const domainWideDelegation = await domainWideDelegationRepository.findByIdIncludeSensitiveServiceAccountKey(
+    {
+      id: credential.delegatedToId,
+    }
+  );
+
+  if (!domainWideDelegation)
+    throw new Error(
+      `Credential with delegatedToId=${credential.delegatedToId} doesn't have a valid domain wide delegation`
+    );
+
+  if (!domainWideDelegation.serviceAccountKey)
+    throw new Error(
+      `Domain wide delegation ${domainWideDelegation.id} doesn't have a valid service account key`
+    );
+
+  return {
+    ...credential,
+    delegatedTo: {
+      serviceAccountKey: domainWideDelegation.serviceAccountKey,
+    },
+  } as CredentialForCalendarService<T>;
+}
+
+export const getCalendarCredentials = async (credentials: Array<CredentialPayload>) => {
+  const calendarCredentials = await Promise.all(
+    getApps(credentials, true)
+      .filter((app) => app.type.endsWith("_calendar"))
+      .flatMap(async (app) => {
+        const credentials = await Promise.all(
+          app.credentials.map(async (credential) => {
+            const credentialForCalendarService = await getCredentialForCalendarService(credential);
```

Comment:
This would be querying DB(via `getCredentialForCalendarService`)  per credential. All credentials would mostly have same `delegatedToId` and thus it is generating the same query unnecessarily per item in the loop.

Though I believe Prisma, would generate just one query for this and reuse the result for others (though not sure). Explicitly ensuring this could be safer.

Reply:
I have a fix for it in a stacked PR.